File: decision_card_system/decision_card_contract.py
from __future__ import annotations

from typing import Literal, NotRequired, TypedDict


DecisionCardContractVersion = Literal["v1"]
DecisionCardState = Literal["generated", "surfaced", "acknowledged", "resolved", "applied"]
DecisionCardEventType = Literal[
    "DecisionCardGenerated",
    "DecisionCardSurfaced",
    "DecisionCardAcknowledged",
    "DecisionCardResolved",
    "DecisionCardApplied",
    "DecisionDeferred",
    "DecisionIgnored",
    "DecisionCompleted",
]


class DecisionCardRecord(TypedDict):
    decision_card_id: str
    household_id: str
    title: str
    root_cause_key: str
    contract_version: DecisionCardContractVersion
    origin_api: str
    dedupe_key: str
    state: DecisionCardState
    created_at: str
    updated_at: str
    actor_id: str
    last_event_type: DecisionCardEventType
    metadata: NotRequired[dict[str, object]]
    resolved_at: NotRequired[str]
    applied_at: NotRequired[str]
    defer_to_date: NotRequired[str]


class DecisionCardEventPayload(TypedDict, total=False):
    event_type: DecisionCardEventType
    decision_card_id: str
    household_id: str
    actor_id: str
    title: str
    root_cause_key: str
    contract_version: DecisionCardContractVersion
    origin_api: str
    dedupe_key: str
    timestamp: str
    metadata: dict[str, object]
    defer_to_date: str
    resolution_kind: str
    decision_id: str

File: hpal-frontend/src/runtime/decision_card_contract.ts
export const DECISION_CARD_CONTRACT_VERSION = "v1" as const;

export type DecisionCardContractVersion = typeof DECISION_CARD_CONTRACT_VERSION;

export type DecisionCardState =
  | "generated"
  | "surfaced"
  | "acknowledged"
  | "resolved"
  | "applied";

export type DecisionCardEventType =
  | "DecisionCardGenerated"
  | "DecisionCardSurfaced"
  | "DecisionCardAcknowledged"
  | "DecisionCardResolved"
  | "DecisionCardApplied"
  | "DecisionDeferred"
  | "DecisionIgnored"
  | "DecisionCompleted";

export interface DecisionCardRecord {
  decision_card_id: string;
  household_id: string;
  title: string;
  root_cause_key: string;
  contract_version: DecisionCardContractVersion;
  origin_api: string;
  dedupe_key: string;
  state: DecisionCardState;
  created_at: string;
  updated_at: string;
  actor_id: string;
  last_event_type: DecisionCardEventType;
  metadata?: Record<string, unknown>;
  resolved_at?: string;
  applied_at?: string;
  defer_to_date?: string;
}
